### PR TITLE
DOSBox-X flatpak support

### PR DESCRIPTION
**Pull request by rderooy**
Flatpak support for the DOSBox-X project, which is a greatly enhanced DOSBox fork.
https://github.com/joncampbell123/dosbox-x

The main developer of DOSBox-X is @joncampbell123 and the largest contributor is @Wengier please add them to this project such that they can make updates.

I have one issue that I don't know how to solve, or if it is even possible. DOSBox-X supports optional NE2000 network emulation, and uses libpcap for that purpose. However for that to work the user needs to have either root permissions, or what we normally suggest is to use setcap to give the necessary permissions to the dosbox-x binary.

e.g.
``setcap cap_net_raw=ep dosbox-x``

This obviously fails as it requires root permissions. I don't know if there is perhaps a solution using polkit?

### CODE REVIEW

**Comment by rderooy** [comment]
bot, build com.dosbox_x.DOSBox-X

File: com.dosbox_x.DOSBox-X.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+sdk: org.freedesktop.Sdk
+
+command: dosbox-x
+
+finish-args:	# flatpak permissions
+  - --device=all	# needed for gamepads and serial/parallel
+  - --share=ipc	        # needed for X11
+  - --share=network	# needed for networking (NE2000, IPX)
+  - --socket=x11
+  - --nosocket=wayland  # not compatible with wayland at this point
```
Comment (bilelmoussaoui) [inline com.dosbox_x.DOSBox-X.yaml:13]:
```suggestion
```

File: com.dosbox_x.DOSBox-X.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      - type: archive
+        url: https://github.com/joncampbell123/dosbox-x/archive/dosbox-x-v0.83.6.tar.gz
+        sha256: 874450dd6f879376ce803984f7614b0edf4eb9d04117c9a88cbb3d60ce69abf0
+    make-args:
+       - -j3
+    post-install:
+      - install -Dm644 /app/share/applications/dosbox-x.desktop /app/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 /app/share/icons/hicolor/scalable/apps/dosbox-x.svg /app/share/icons/hicolor/scalable/apps/${FLATPAK_ID}.svg
+      - install -Dm644 /app/share/metainfo/dosbox-x.metainfo.xml /app/share/metainfo/${FLATPAK_ID}.metainfo.xml
+      - sed -i s/Icon=.*/Icon=${FLATPAK_ID}/ /app/share/applications/${FLATPAK_ID}.desktop
```
Comment (bilelmoussaoui) [inline com.dosbox_x.DOSBox-X.yaml:61]: Use desktop-file-edit

File: com.dosbox_x.DOSBox-X.yaml (hunk 1 above)
Comment (rderooy) [reply com.dosbox_x.DOSBox-X.yaml:13]: As the comment clearly says, wayland does not work at this point.

Comment (bilelmoussaoui) [reply com.dosbox_x.DOSBox-X.yaml:13]: Well --nosocket=wayland is what you have by default.

**Comment by rderooy** [comment]
bot, build com.dosbox_x.DOSBox-X

File: com.dosbox_x.DOSBox-X.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    config-opts:
+      - --enable-core-inline
+      - --enable-debug=heavy
+      - --enable-sdl2
+    sources:
+      - type: archive
+        url: https://github.com/joncampbell123/dosbox-x/archive/dosbox-x-v0.83.6.tar.gz
+        sha256: 874450dd6f879376ce803984f7614b0edf4eb9d04117c9a88cbb3d60ce69abf0
+    make-args:
+       - -j3
```
Comment (bilelmoussaoui) [inline com.dosbox_x.DOSBox-X.yaml:55]: Flatpak-builder already uses the `$FLATPAK_BUILDER_N_JOBS` value by default.
```suggestion
```

**Comment by rderooy** [comment]
bot, build com.dosbox_x.DOSBox-X

File: com.dosbox_x.DOSBox-X.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+runtime-version: '20.08'
+sdk: org.freedesktop.Sdk
+
+command: dosbox-x
+
+finish-args:	# flatpak permissions
+  - --device=all	# needed for gamepads and serial/parallel
+  - --share=ipc	        # needed for X11
+  - --share=network	# needed for networking (NE2000, IPX)
+  - --socket=x11
```
Comment (barthalion) [inline com.dosbox_x.DOSBox-X.yaml:12]: Doesn't it work natively on Wayland due to SDL2?

Comment (barthalion) [reply com.dosbox_x.DOSBox-X.yaml:12]: Just tested and it seems fine, should be replaced with `--socket=wayland` and `--socket=fallback-x11`.

**Comment by barthalion** [comment]
> However for that to work the user needs to have either root permissions, or what we normally suggest is to use setcap to give the necessary permissions to the dosbox-x binary.

I don't think bubblewrap allows that so I guess it means libpcap can be dropped?

File: com.dosbox_x.DOSBox-X.yaml (hunk 4 above)
Comment (rderooy) [reply com.dosbox_x.DOSBox-X.yaml:12]: Unfortunately it is not fine. I tested it here on Fedora 33 beta and launching with wayland causes the window decorations and controls to vanish. In addition the window is the wrong size and the mouse location tracking gets messed up. I think the later two issues are due to the fact that the title bar is missing and the window gets stretched to the window+title-bar which in turn causes the mouse position to get messed up.

**Comment by rderooy** [comment]
> > However for that to work the user needs to have either root permissions, or what we normally suggest is to use setcap to give the necessary permissions to the dosbox-x binary.
> 
> I don't think bubblewrap allows that so I guess it means libpcap can be dropped?

Yes if there is no way to allow the necessary permissions than there is no point including libpcap. A secondary solution is being worked on using libslirp which does not require root permissions and will hopefully work in flatpak. Even though libslirp brings along it's own set of restrictions (e.g. libslirp only works with TCP/IP, while libpcap allows any legacy protocol).

**Comment by rderooy** [comment]
bot, build com.dosbox_x.DOSBox-X

File: com.dosbox_x.DOSBox-X.yaml (hunk 4 above)
Comment (rderooy) [reply com.dosbox_x.DOSBox-X.yaml:12]: This issue is also not specific to flatpak. The same thing happens when running the binary normally using
```
SDL_VIDEODRIVER=wayland dosbox-x
```

Comment (barthalion) [reply com.dosbox_x.DOSBox-X.yaml:12]: Ack, please ignore my comment.

**Pull request closed without merging**